### DISCUSSION TITLE
Could not find field 'RowNum' on row of type 'TblPurchaseDetailsRow' ?

### DISCUSSION THREAD

**Discussion by ra-bedo11** [serenity-is/Serenity#6181] [Help] [answered]
![error_Save_Details1](https://user-images.githubusercontent.com/29665710/148465552-ec06e9b5-5b89-4caa-9cd6-d3794712da52.jpg)

and This My Code in grid.ts

/// <reference path="../../Common/Helpers/GridEditorBase.ts" />

namespace Serene3.Default {

    import fld = Default.TblPurchaseDetailsRow.Fields;

    //@Serenity.Decorators.registerClass()
    @Serenity.Decorators.registerEditor()

    //export class TblPurchaseDetailsGrid extends Serenity.EntityGrid<TblPurchaseDetailsRow, any> {
    export class TblPurchaseDetailsGrid extends Common.GridEditorBase<TblPurchaseDetailsRow> {

        protected getColumnsKey() { return 'Default.TblPurchaseDetails'; }
        protected getDialogType() { return TblPurchaseDetailsDialog; }
        //protected getIdProperty() { return TblPurchaseDetailsRow.idProperty; }
        //protected getInsertPermission() { return TblPurchaseDetailsRow.insertPermission; }
        protected getLocalTextPrefix() { return TblPurchaseDetailsRow.localTextPrefix; }
        //protected getService() { return TblPurchaseDetailsService.baseUrl; }

        protected get_ExtGridOptions(): ExtGridOptions { return Q.deepClone(q.DefaultMainGridOptions); }

        constructor(container: JQuery) {
            super(container);

            container.css({ minHeight: 280 })

            new _Ext.HeaderFiltersMixin({
                grid: this
            })
        }

        protected getAddButtonCaption() {
            return "Choose Product";
        }

        protected validateEntity(row: TblPurchaseDetailsRow, id: number) {
            if (!super.validateEntity(row, id))
                return false;
            row.ProductProductNameAr = TblProductRow.getLookup().itemById[row.ProductId].ProductNameAr;
            row.UnitUnitNameAr = TblUnitRow.getLookup().itemById[row.UnitId].UnitNameAr;

            var sameProduct = Q.tryFirst(this.view.getItems(), x => x.ProductId === row.ProductId);

            if (sameProduct && this.id(sameProduct) !== id) {
                Q.alert('This product is already Please Choose Another!');
                return false;
            }

            return true;
        }

        //public getGriditems() {
        //    super.getGrid();
        //    var items = "select * from events where departmentid=this.form.departmentid";
        //    return items;
        //}

        protected createSlickGrid(): Slick.Grid {
            var grid = super.createSlickGrid();
            grid.setSelectionModel(new Slick.RowSelectionModel());

            // need to register this plugin for grouping or you'll have errors
            grid.registerPlugin(new Slick.Data.GroupItemMetadataProvider());

            this.view.setSummaryOptions({
                aggregators: [
                    new Slick.Aggregators.Sum(fld.TotalBill),
                    new Slick.Aggregators.Sum(fld.PurchaseQte),
                    new Slick.Aggregators.Sum(fld.Total)

                ]
            });
            return grid;
        }

        protected getSlickOptions(): Slick.GridOptions {
            var opt = super.getSlickOptions();
            opt.enableTextSelectionOnCells = true;
            opt.editable = true;
            opt.selectedCellCssClass = "slick-row-selected";
            opt.enableCellNavigation = true;

            opt.showFooterRow = true;
            return opt;
        }

        nextRowNumber = 1;

        protected getColumns(): Slick.Column[] {
            let columns = super.getColumns();

            let extOptions = this.get_ExtGridOptions();

            columns.unshift({
                field: 'RowNum',
                name: '#',
                cssClass: 'rownum-column',
                headerCssClass: 'align-center',
                width: 40,
                minWidth: 40,
                maxWidth: 40,
                visible: extOptions.ShowRowNumberColumn,
                format: ctx => {
                    if (!ctx.item.RowNum) {
                        ctx.item.RowNum = this.nextRowNumber++;
                    }
                    return String(ctx.item.RowNum);
                }
            });

            return columns;
        }

        protected resetRowNumber() {
            this.nextRowNumber = 1;
            let items = this.getItems();

            let grouping_fields = this.view.getGrouping();

            if (grouping_fields.length == 0) {
                for (let i = 0; i < items.length; i++) {
                    (items[i] as any).RowNum = i + 1;
                }
            } else if (grouping_fields.length > 0) {

                let generateRowNumber = (groups: Slick.Group<any>[]) => {

                    for (let gi = 0; gi < groups.length; gi++) {
                        let subGroups = groups[gi].groups;
                        if (subGroups) {
                            generateRowNumber(subGroups);
                        } else {
                            let rows = groups[gi].rows;
                            for (let i = 0; i < rows.length; i++) {
                                rows[i].RowNum = i + 1;
                            }
                        }
                    }
                };

                let groups = this.view.getGroups();
                generateRowNumber(groups);
            }

            this.setItems(items);
        }

        protected onViewProcessData(response: Serenity.ListResponse<TblPurchaseDetailsRow>): Serenity.ListResponse<TblPurchaseDetailsRow> {
            let r = super.onViewProcessData(response);

            if (this.get_ExtGridOptions().ShowRowNumberColumn == true) {
                setTimeout(() => { this.resetRowNumber() });
            }
            return r;
        }

        //protected onViewProcessData(response: Serenity.ListResponse<TblPurchaseDetailsRow>): Serenity.ListResponse<TblPurchaseDetailsRow> {
        //    response = super.onViewProcessData(response);

        //    for (var x of response.Entities) {
        //        (x as any).__id = this.nextRowNumber++;
        //    }
        //    if (this.get_ExtGridOptions().ShowRowNumberColumn == true) {
        //        setTimeout(() => { this.resetRowNumber() });
        //    }
        //    return response;
        //}
    }
}

**Comment by wezmag** [comment, marked as answer]
In your `TblPurchaseDetailsRow.cs` add a field called `RowNum` and use `NotMapped` attribute.

```
[NotMapped]
public int? RowNum
{
     get => fields.RowNum[this];
     set => fields.RowNum[this] = value;
}
```

**Reply by ra-bedo11** [reply]
Thanks
